crates/kubelet/src/server.rs

```diff
@@ -1,189 +1,113 @@
 /// Server is an HTTP(S) server for answering Kubelet callbacks.
 ///
 /// Logs and exec calls are the main things that a server should handle.
-use anyhow::Context;
-use hyper::service::service_fn;
-use hyper::{server::conn::Http, Body, Method, Request, Response, StatusCode};
-use log::{debug, error, info, warn};
-use native_tls::{Identity, TlsAcceptor};
-use tokio::net::{TcpListener, TcpStream};
-use tokio::stream::StreamExt;
-
+use log::{debug, error};
+use std::convert::Infallible;
 use std::sync::Arc;
+use warp::Filter;
 
-use crate::config::ServerConfig;
-use crate::logs::LogSender;
-use crate::provider::{NotImplementedError, Provider};
-
-/// Start the Krustlet HTTP(S) server
-///
+/// Start the Krustlet HTTP(S) server                                                                                                                                                                                                                       │
+///                                                                                                                                                                                                                                                         │
 /// This is a primitive implementation of an HTTP provider for the internal API.
-/// TODO: Support TLS/SSL.
 pub async fn start_webserver<T: 'static + Provider + Send + Sync>(
     provider: Arc<T>,
     config: &ServerConfig,
 ) -> anyhow::Result<()> {
-    let identity = tokio::fs::read(&config.pfx_path)
-        .await
-        .with_context(|| format!("Could not read file {:?}", config.pfx_path))?;
-    let identity = Identity::from_pkcs12(&identity, &config.pfx_password)?;
-
-    let acceptor = tokio_tls::TlsAcceptor::from(TlsAcceptor::new(identity)?);
-
-    let acceptor = Arc::new(acceptor);
-
-    let address = std::net::SocketAddr::new(config.addr, config.port);
-    let mut listener = TcpListener::bind(&address).await.unwrap();
-
-    info!("Starting webserver at: {}", address);
-
-    let mut incoming = listener.incoming();
-
-    while let Some(conn) = incoming.try_next().await? {
-        let acceptor = acceptor.clone();
-        let provider = provider.clone();
-
-        tokio::spawn(async move {
-            if let Err(e) = handle_connection(conn, acceptor, provider).await {
-                error!("Error handling server connection: {}", e);
-            }
+    let health = warp::get().and(warp::path("healthz")).map(get_ping);
+    let ping = warp::get().and(warp::path::end()).map(get_ping);
+
+    let logs_provider = provider.clone();
+    let logs = warp::get()
+        .and(warp::path!("containerLogs" / String / String / String))
+        .and(warp::query::<LogOptions>())
+        .and_then(move |namespace, pod, container, opts| {
+            let provider = logs_provider.clone();
+            get_container_logs(provider, namespace, pod, container, opts)
         });
-    }
 
-    Ok(())
-}
+    let exec_provider = provider.clone();
```

Comment:
I don't believe we need this extra clone. Do we?

Reply:
I'm not sure if there is a cleaner way to do this, but both closures need to move a provider.